src/Jackalope/Version/VersionHandler.php

```diff
@@ -235,15 +236,199 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+        }
+
+        // handle properties present on the frozen node
+        foreach ($frozenNode->getProperties() as $property) {
+            /** @var PropertyInterface $property */
+            $propertyName = $property->getName();
+            if ($propertyName == 'jcr:frozenPrimaryType'
+                || $propertyName == 'jcr:frozenMixinTypes'
+                || $propertyName == 'jcr:frozenUuid'
+            ) {
+                continue;
+            }
+
+            if ($node->hasProperty($propertyName)) {
+                $nodeProperty = $node->getProperty($propertyName);
+                $onParentVersion = $nodeProperty->getDefinition()->getOnParentVersion();
+                if ($onParentVersion == OnParentVersionAction::COPY || $onParentVersion == OnParentVersionAction::VERSION) {
+                    $nodeProperty->setValue($property->getValue());
+                }
+            } else {
+                // cannot check the onParentVersion attribute of a non existing property
+                // but if the property exists on the frozen node it has to be copy or version
+                $node->setProperty($propertyName, $property->getValue());
+            }
+        }
+
+        // handle properties present on the node but not on the frozen node
+        foreach ($node->getProperties() as $propertyName => $property) {
+            if ($frozenNode->hasProperty($propertyName)) {
+                continue;
+            }
+
+            $onParentVersion = $property->getDefinition()->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $property->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                case OnParentVersionAction::COMPUTE: // implementation-specific, decided to handle as INITIALIZE
+                    // TODO how to get default value of property?
+            }
+        }
+
+        // TODO handle identifier collisions
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            $childNodePath = $node->getPath() . '/' . $frozenChildNode->getName();
+            if ($this->session->nodeExists($childNodePath)) {
+                $this->session->removeItem($childNodePath);
+            }
+
+            if (!$frozenChildNode->isNodeType('nt:versionedChild')) {
+                $this->restoreFromNode($node, $frozenChildNode);
+            } else {
+                // TODO handle restoring from frozenChildNode with OPV of VERSION
```

Comment:
I can't imagine that a lot of system want to rely on that anyway, because the specification is not really clear about what happens here:

> Under full versioning each child node C of N where C has an OPV of VERSION and C is versionable, is represented in F not as a copy of N/C but as special node of type nt:versionedChild containing a reference to the version history of C. On restore, N/C in the workspace is replaced by a version of C. The determination of which version of C to use is implementation-dependent (see §15.7.5 Chained Versions on Restore).

(See 15.7.6 on https://docs.adobe.com/content/docs/en/spec/jcr/2.0/15_Versioning.html)